File: python/0012.py
class Solution(object):
    def intToRoman(self, num):
        """
        :type num: int
        :rtype: str
        """
        ma={
            'I':1,
            'IV':4,
            'V':5,
            'IX':9,
            'X':10,
            'XL':40,
            'L':50,
            'XC':90,
            'C':100,
            'CD':400,
            'D':500,
            'CM':900,
            'M':1000,
        }
        ans=''
        if num>=1000:

            ans+='M'*(num//1000)
            num%=1000
        if num>=900:
            ans+='CM'
            num%=100
        if num>=500:
            ans+='D'+'C'*(num//100-5)
            num%=100
        if num>=400:
            ans+='CD'
            num%=100
        if num>=100:
            ans+='C'*(num//100)
            num%=100
        if num>=90:
            ans+="XC"
            num%=10
        if num>=50:
            ans+='L'+'X'*(num//10-5)
            num%=10
        if num>=40:
            ans+="XL"
            num%=10
        if num>=10:
            ans+='X'*(num//10)
            num%=10
        if num>=9:
            ans+='IX'
            num=0
        if num>=5:
            ans+='V'+'I'*(num-5)
            num=0
        if num>=4:
            ans+='IV'
            num=0
        if num>0:
            ans+='I'*num
            num=0
        return ans

File: rust/src/bin/0012.rs
use std::collections::HashMap;
use std::ops::{Add, AddAssign};

struct Solution;

fn get(n: i32, one: char, five: char, ten: char) -> String {
    if n <= 3 {
        return one.to_string().repeat(n as usize);
    } else if n == 4 {
        let mut x = one.to_string();
        x.push(five);
        return x;
    } else if n < 9 {
        let mut x = five.to_string();
        for i in 6..=n {
            x.push(one);
        }
        return x;
    } else if n == 9 {
        let mut s = one.to_string();
        s.push(ten);
        return s;
    }
    unreachable!()
}

impl Solution {
    pub fn int_to_roman(num: i32) -> String {
        let mut x = num;
        let m = [
            ("I", 1),
            ("V", 5),
            ("X", 10),
            ("L", 50),
            ("C", 100),
            ("D", 500),
            ("M", 1000),
        ];
        let mut ans = "".to_string();
        ans.add_assign(&get(x / 1000, 'M', 'Z', 'W'));
        x %= 1000;
        ans.add_assign(&get(x / 100, 'C', 'D', 'M'));
        x %= 100;
        ans.add_assign(&get(x / 10, 'X', 'L', 'C'));
        x %= 10;
        ans.add_assign(&get(x, 'I', 'V', 'X'));
        ans
    }
}

fn main() {
    println!("{}", Solution::int_to_roman(3));
}